### ISSUE TITLE
Add better control of inclusion/exclusion of certain file/buf/...-types 

### ISSUE THREAD

**Issue by Danielkonge** [lukas-reineke/indent-blankline.nvim#778]
### Problem

I would like to use `indent-blankline` in `InspectTree`, but `InspectTree` is of buftype `'nofile'`. Of course I could just remove `'nofile'` from the excluded buftypes, but I do actually want to exclude `'nofile'` in all other cases. Therefore it would be nice to have slightly better control of whether to use `indent-blankline` in certain cases. 

### Expected behavior

I am not sure what the best solution to this would be with the way `indent-blankline` works right now. 

In `rainbow-delimiters` I have done something slightly similar in highlighting of `query` files, where I highlight one way if `vim.b.dev_base` is truthy (i.e., we are in `InspectTree`) and another way if `vim.b.dev_base` is `nil`. I can do this here by allowing setup via a function that I then use the return value of. 

If you have an idea for how to do something like I want, I can implement a draft of it, if you would want that. 

Alternatively, is there already a way to do what I want that I missed?

**Comment by Danielkonge** [comment]
Update: I tried removing `'nofile'` from the exluded buftypes and adding the following hook:
```lua
hooks.register(
    hooks.type.ACTIVE,
    function(_)
        if vim.bo.buftype == 'nofile' then
            if vim.b.dev_base then
                return true
            else
                return false
            end
        end
        return true
    end
)
```
But this didn't work. However when I go and edit out `'nofile'` from the exclude list in the source code, then it works, so it seems there is still some problem with updating the config correctly. I will try to find the problem.

**Comment by lukas-reineke** [comment]
> Add a way to overwrite the excludes.

There is, it's called `:help ibl.overwrite()`. With this, you can completely replace excludes instead of adding to it. Then your hook should work fine.

**Comment by Danielkonge** [comment]
`ibl.overwrite()` didn't give me the behavior I want. I think it is because of this line:
https://github.com/lukas-reineke/indent-blankline.nvim/blob/29be0919b91fb59eca9e90690d76014233392bef/lua/ibl/config.lua#L274

I changed
```lua
    M.config = merge_configs("merge", M.default_config, config)
```
into
```lua
    M.config = merge_configs("merge", M.config or M.default_config, config)
```
and now I get the expected behavior of `ibl.overwrite()` using 
```lua
ibl.overwrite({
    exclude = {
        buftypes = {
            "terminal",
            -- don't include "nofile",
            "quickfix",
            "prompt",
        }
    }
})
```
and my above hook.

**Comment by Danielkonge** [comment]
I made a pull request #779 with this one line code change that fixes my issue.

**Comment by lukas-reineke** [comment]
Can you explain again what problem you have with `overwrite`?

For me, this works as expected.

```lua
local lazypath = "/tmp/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
    vim.fn.system { "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath }
end
vim.opt.rtp:prepend(lazypath)
vim.opt.termguicolors = true

require("lazy").setup ({
    {
        "lukas-reineke/indent-blankline.nvim",
        config = function()
            local ibl = require("ibl")
            ibl.setup()
            ibl.overwrite {
                exclude = {
                    buftypes = {
                        "terminal",
                        -- don't include "nofile",
                        "quickfix",
                        "prompt",
                    }
                }
            }
        end
    },
}, { root = "/tmp/lazy" })
```

**Comment by Danielkonge** [comment]
Since my pull request was not the correct fix let me try to clarify my issue. Here is a simplified version of my setup of `ibl`:

```lua
local hooks = require "ibl.hooks"
hooks.register(
    hooks.type.ACTIVE,
    function(_)
        if vim.bo.buftype == 'nofile' then
            if vim.b.dev_base then
                -- we are in `InspectTree`
                return true
            else
                return false
            end
        end
        return true
    end
)

local ibl = require("ibl")
ibl.overwrite({
    exclude = {
        buftypes = {
            "terminal",
            -- don't include "nofile",
            "quickfix",
            "prompt",
        }
    }
})
ibl.setup({
    --
})
```


The behavior I would expect from this is the standard behavior of the plugin with one change: I wan't indentation guides in `:InspectTree<CR>` but not in other buffer of buftype `'nofile'`.

With the current code this doesn't work. I can see in your answer to my pull request that `ibl.overwrite()` shouldn't be affected by a change to that line of code, but it does change the behavior when I test it. 

With the current code I don't get the expected behavior, but with that one line change in the code I do get the expected behavior. (I also get the expected behavior if I instead manually remove `"nofile"` from the config in a fork.)

**Comment by Danielkonge** [comment]
> Can you explain again what problem you have with `overwrite`?
> 
> For me, this works as expected.
> 
> ```lua
> local lazypath = "/tmp/lazy/lazy.nvim"
> if not vim.loop.fs_stat(lazypath) then
>     vim.fn.system { "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath }
> end
> vim.opt.rtp:prepend(lazypath)
> vim.opt.termguicolors = true
> 
> require("lazy").setup ({
>     {
>         "lukas-reineke/indent-blankline.nvim",
>         config = function()
>             local ibl = require("ibl")
>             ibl.setup() -- your normal config
>             ibl.overwrite {
>                 exclude = {
>                     buftypes = {
>                         "terminal",
>                         -- don't include "nofile",
>                         "quickfix",
>                         "prompt",
>                     }
>                 }
>             }
>         end
>     },
> }, { root = "/tmp/lazy" })
> ```

Let me test it a bit more. It might be the hook not workin as expected based on this.

**Comment by Danielkonge** [comment]
No wait. I think it is just me calling the functions in the wrong order actually.

**Comment by lukas-reineke** [comment]
Yes, you just need to call setup first. Setup will reset everything you do before.

**Comment by Danielkonge** [comment]
> Yes, you just need to call setup first. Setup will reset everything you do before.

Thank you for the help with finding the problem. I think it works as expected now. I do think it would make sense to add a note about calling overwrite after setup to the docs though.
